src/recon_buildblock/CUDA_stir/CudaRelativeDifferencePrior.cu

```diff
@@ -0,0 +1,382 @@
+    }
+
+    if (this->weights.get_length() == 0)
+    {
+        compute_weights(this->weights, current_image_cast.get_grid_spacing(), this->only_2D);
+    }
+    auto kappa_ptr = this->get_kappa_sptr();
+    const bool do_kappa = !is_null_ptr(kappa_ptr);
+    if (do_kappa && !kappa_ptr->has_same_characteristics(current_image_estimate))
+        error("RelativeDifferencePrior: kappa image has not the same index range as the reconstructed image\n");
+
+    const int z_dim = this->z_dim;
+    const int y_dim = this->y_dim;
+    const int x_dim = this->x_dim;
+
+    std::vector<float> image_data(current_image_estimate.size_all());
+    std::vector<float> weights_data(this->weights.size_all());
+    std::copy(current_image_estimate.begin_all(), current_image_estimate.end_all(), image_data.begin());
+    std::copy(this->weights.begin_all(), this->weights.end_all(), weights_data.begin());
+
+    float *d_image_data, *d_weights_data, *d_gradient_data;
+
+    // Allocate memory on the GPU
+    cudaMalloc(&d_image_data, current_image_estimate.size_all() * sizeof(float));
+    cudaMalloc(&d_weights_data, this->weights.size_all() * sizeof(float));
+    cudaMalloc(&d_gradient_data, prior_gradient.size_all() * sizeof(float));
+    
+    // Copy data from host to device
+    cudaMemcpy(d_image_data, image_data.data(), current_image_estimate.size_all() * sizeof(float), cudaMemcpyHostToDevice);
+    cudaMemcpy(d_weights_data, weights_data.data(), this->weights.size_all() * sizeof(float), cudaMemcpyHostToDevice);
+
+    if (do_kappa)
+    {
+        std::vector<float> kappa_data(current_image_estimate.size_all());
+        std::copy(kappa_ptr->begin_all(), kappa_ptr->end_all(), kappa_data.begin());
+        float *d_kappa;
+        cudaMalloc(&d_kappa, current_image_estimate.size_all() * sizeof(float));
+        cudaMemcpy(d_kappa, kappa_data.data(), current_image_estimate.size_all() * sizeof(float), cudaMemcpyHostToDevice);
+        computeCudaRelativeDifferencePriorGradientKernel<<<this->grid_dim, this->block_dim>>>(
+            d_gradient_data, d_image_data, d_weights_data, d_kappa, do_kappa, this->gamma, this->epsilon, z_dim, y_dim, x_dim
+        );
+        cudaFree(d_kappa);
+    }
+    else
+    {
+        computeCudaRelativeDifferencePriorGradientKernel<<<this->grid_dim, this->block_dim>>>(
+            d_gradient_data, d_image_data, d_weights_data, nullptr, do_kappa, this->gamma, this->epsilon, z_dim, y_dim, x_dim
+        );
+    }
+
+    // Check for any errors during kernel execution
+    cudaError_t error = cudaGetLastError();
+    if (error != cudaSuccess) {
+        std::cerr << "CUDA error in compute_value kernel execution: " << cudaGetErrorString(error) << std::endl;
+        cudaFree(d_image_data); cudaFree(d_weights_data); cudaFree(d_gradient_data);
+    }
+
+    // Allocate host memory for the result and copy from device to host
+    std::vector<float> gradient_data(current_image_estimate.size_all());
```

Comment:
ok, but could reuse some memory (might be better to leave this though for clarity and future changes)